src/messaging/ExchangeDelegate.h

```diff
@@ -91,7 +91,16 @@ class DLL_EXPORT ExchangeDelegate
     /**
      * @brief
      *   This function is the protocol callback to invoke when the associated
-     *   exchange context is being closed
+     *   exchange context is being closed.
+     *
+     *   If the exchange was in a state where it was expecting a message to be
+     *   sent due to an earlier WillSendMessage call or because the exchange has
+     *   just been created as an initiator, the consumer is holding a reference
+     *   to the exchange and it's the consumer's responsibility to call
+     *   Release() on the exchange at some point.  The usual way this happens is
+     *   that the consumer tries to send its message, that fails, and the
+     *   consumer calls Close() on the exchange.  Calling Close() after an
+     *   OnExchangeClosing() notification is allowed.
```

Comment:
This is far from true. If a `Close` is called somewhere (e.g. By `MessageHandled`), then the user will receive a `OnExchangeClosing` event, if the user calls `Close` again, it will decrease the ref again, and can make it negative.

We should disallow the user doing anything to that exchange after `OnExchangeClosing` is triggered.

Reply:
>  If a Close is called somewhere (e.g. By MessageHandled)

Then the exchange is not in a state where it was waiting to send a message, right?  Maybe I should make it clearer: calling Close() after OnExchangeClosing() is allowed _if the exchange was in waiting for send state_ when it closed.